beacon-chain/forkchoice/protoarray/store.go

```diff
@@ -578,52 +578,42 @@ func (s *Store) prune(ctx context.Context, finalizedRoot [32]byte) error {
 		return nil
 	}
 
-	// Remove the key/values from indices mapping on to be pruned nodes.
-	// These nodes are before the finalized index.
-	for i := uint64(0); i < finalizedIndex; i++ {
-		if int(i) >= len(s.nodes) {
-			return errInvalidNodeIndex
+	// We start adding nodes since the final one, anything that arrived
+	// to fork choice before the finalized block is either final now or
+	// it is in a contentious fork.
+	processedNodes := make(map[uint64]uint64, uint64(len(s.nodes))-finalizedIndex)
+	newNodes := make([]*Node, 1, uint64(len(s.nodes))-finalizedIndex)
+	node := s.nodes[finalizedIndex]
+	node.parent = NonExistentNode
+	newNodes[0] = node
+	processedNodes[finalizedIndex] = uint64(0)
+	for idx := uint64(0); idx < uint64(len(s.nodes)); idx++ {
+		node = s.nodes[idx]
+		parentIdx, ok := processedNodes[node.parent]
+		if ok {
+			// This node is a descendant of the finalized point
+			s.nodesIndices[node.root] = uint64(len(newNodes))
+			processedNodes[idx] = uint64(len(newNodes))
+			node.parent = parentIdx
+			newNodes = append(newNodes, node)
+		} else {
+			// This node is in a contentious fork
+			delete(s.nodesIndices, node.root)
 		}
-		delete(s.nodesIndices, s.nodes[i].root)
```

Comment:
I think we still need to clean  `s.nodesIndices` where `i < finalizedIndex`

Reply:
Nodes with `i < finalizedIndex` cannot be descendant of the finalized node, therefore they are deleted in the previous loop. I can update the comment "This node is in a contentious fork" to "This node is not a descentant of the finalized state"

Reply:
Ah, i see it. Neat!